src/core/utils/slaac_address.cpp

```diff
@@ -173,7 +173,9 @@ void Slaac::Update(UpdateMode aMode)
                         continue;
                     }
 
-                    if (config.mSlaac && !ShouldFilter(prefix) && (prefix.mLength == slaacAddr->mPrefixLength) &&
+                    if (config.mSlaac && !ShouldFilter(prefix) &&
+                        ((config.mOnMesh && (slaacAddr->mPrefixLength == prefix.mLength)) ||
+                         (!config.mOnMesh && (slaacAddr->mPrefixLength == 128))) &&
```

Comment:
Define a new method like `OnMeshPrefixConfig::Match` to check if the address is matched by the Prefix to encapsulate this condition?